crates/edr_eth/src/signature/fakeable.rs

```diff
@@ -0,0 +1,217 @@
+    }
+
+    /// Whether the signature is from an impersonated account.
+    pub fn is_fake(&self) -> bool {
+        matches!(self.data, FakeableData::Fake { .. })
+    }
+
+    /// Recovers the Ethereum address which was used to sign the transaction.
+    pub fn recover_address<MessageT>(&self, message: MessageT) -> Result<&Address, SignatureError>
+    where
+        MessageT: Into<RecoveryMessage>,
+    {
+        self.address.get_or_try_init(|| {
+            let message = message.into();
+
+            match &self.data {
+                FakeableData::Recoverable { signature } => signature.recover_address(message),
+                FakeableData::Fake { .. } => {
+                    unreachable!("fake signature must be initialized with address")
+                }
+            }
+        })
+    }
+}
+
+// We always assume that a decoded signature is a recoverable signature.
+impl<SignatureT> alloy_rlp::Decodable for Fakeable<SignatureT>
+where
+    SignatureT: alloy_rlp::Decodable + Recoverable + Signature,
+{
+    fn decode(buf: &mut &[u8]) -> alloy_rlp::Result<Self> {
+        let signature = SignatureT::decode(buf)?;
+
+        Ok(Self {
+            data: FakeableData::Recoverable { signature },
+            address: OnceCell::new(),
+        })
+    }
+}
+
+// We need a custom implementation to avoid the struct being treated as an RLP
+// list.
+impl<SignatureT: alloy_rlp::Encodable + Recoverable + Signature> alloy_rlp::Encodable
+    for Fakeable<SignatureT>
+{
+    fn encode(&self, out: &mut dyn BufMut) {
+        match &self.data {
+            FakeableData::Fake { v } => {
+                if let Some(y_parity) = self.y_parity() {
+                    EcdsaWithYParity {
+                        r: self.r(),
+                        s: self.s(),
+                        y_parity,
+                    }
+                    .encode(out);
+                } else {
+                    let ecdsa = Ecdsa {
+                        r: self.r(),
+                        s: self.s(),
+                        v: *v,
+                    };
+
+                    println!("ecdsa: {ecdsa:?}");
+
+                    ecdsa.encode(out);
+                }
+            }
+            FakeableData::Recoverable { signature } => signature.encode(out),
+        }
+    }
+
+    fn length(&self) -> usize {
+        match &self.data {
+            FakeableData::Fake { v } => {
+                let v_length = if let Some(y_parity) = self.y_parity() {
+                    y_parity.length()
+                } else {
+                    v.length()
+                };
+
+                self.r().length() + self.s().length() + v_length
+            }
+            FakeableData::Recoverable { signature } => signature.length(),
+        }
+    }
+}
+
+impl<SignatureT: Recoverable + Signature> PartialEq for Fakeable<SignatureT> {
+    fn eq(&self, other: &Self) -> bool {
+        self.r() == other.r() && self.s() == other.s() && self.v() == other.v()
```

Comment:
What's the reason for the custom implementaton?

Reply:
As the address is cached in a `OnceCell` that should not be incorporated in the comparison. Validating the actual signature suffices. I achieved that by comparing the `r`, `s,` and `v` values.

You pointing this out, made me realise that calculating those values has some overhead, so I separated the comparison of `Fake` and `Recoverable` in [this commit](https://github.com/NomicFoundation/edr/pull/498/commits/33d5df811d21864d06cb974daff96bc2d91c7e35).